### PR TITLE
Make the startup lock message print to stderr.

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This changes the "Waiting for another flutter command to release the startup lock..." message output so that it appears on stderr instead of stdout.  When it appears on stdout, it can mess up collection of the output. For instance, if you run `flutter --version --machine` and you're expecting JSON output, then you'll get non-JSON output even though the lock is released and you eventually would get what you're asking for.

## Tests

- Added an integration test that runs a separate process that locks the lock file.

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
What about an integration test? In theory you can start two flutter commands immediately and you have a pretty good chance of hitting the lock message on one. Not sure if that would be reliable enough.

If that doesn't work, a test that uses the real file system would be ideal

**Comment by gspencergoog** [comment]
I'll see if I can make one with the real filesystem, I feel like starting two commands might be flaky.

I think I can just open the lock file and lock it before calling `lock` and maybe that'll work.

**Comment by gspencergoog** [comment]
OK, I've tried a couple of approaches, but neither works.

I tried to make a test with the real filesystem, but the problem there is that locks are per process, so even running it in an isolate won't result in the lock being denied: the process already has a lock on that file, so it says "Sure, you can have the lock: you have it already".

I then tried to run an external process that would lock a file, and while that does lock the file, it means I need to know where the Dart executable is, so that I can be assured that the lock is the same kind of lock (e.g. the obvious alternative of `flock` doesn't work).

Here's what the test looks like (note the absolute path to Dart):
```dart
    testWithoutContext('should log a message to stderr when lock is not acquired', () async {
      final String oldRoot = Cache.flutterRoot;
      final FileSystem fileSystem = LocalFileSystem.test(signals: Signals.test());
      final Directory tempDir = fileSystem.systemTempDirectory.createTempSync('cache_test.');
      final FakeLogger logger = FakeLogger();
      try {
        Cache.flutterRoot = tempDir.absolute.path;
        final Cache cache = Cache.test(
            fileSystem: fileSystem, processManager: FakeProcessManager.any(), logger: logger);
        final File cacheFile = fileSystem.file(fileSystem.path.join(Cache.flutterRoot, 'bin', 'cache', 'lockfile'))
          ..createSync(recursive: true);
        final File script = fileSystem.file(fileSystem.path.join(Cache.flutterRoot, 'bin', 'cache', 'test_lock.sh'));
        script.writeAsStringSync(r'''
import 'dart:async';
import 'dart:io';

Future<void> main(List<String> args) async {
  File file = File(args[0]);
  RandomAccessFile lock = file.openSync(mode: FileMode.write);
  lock.lockSync();
  await Future<void>.delayed(const Duration(milliseconds: 1000));
  exit(0);
}
''');
        final Process process = await const LocalProcessManager().start(<String>['/bin/bash', '-c', '/home/gspencer/code/flutter/bin/cache/dart-sdk/bin/dart ${script.absolute.path} ${cacheFile.absolute.path} || echo Subprocess Failed']);
        await Future<void>.delayed(const Duration(milliseconds: 500));
        await cache.lock();
        process.kill(io.ProcessSignal.sigkill);
      } finally {
        tempDir.deleteSync(recursive: true);
        Cache.flutterRoot = oldRoot;
      }
      expect(logger.errors.single, equals('Waiting for another flutter command to release the startup lock...'));
      expect(logger.status, isEmpty);
    });
```

Do you know of a good way to find out where the Dart executable is in a tools test?

**Comment by jonahwilliams** [comment]
Something like https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/tool_backend_test.dart#L14 ?

**Comment by gspencergoog** [comment]
Yes, that should work.  I've added the test: hopefully it'll work on the server where `FLUTTER_ROOT` is set. Running it locally fails with:

```
Bad state: Could not determine flutter_tools/ path from script URL (file:///tmp/dart_test.VRABED/test.dart_1.dill); consider setting FLUTTER_ROOT explicitly.
```

File: packages/flutter_tools/lib/src/cache.dart [hunk 1]
```diff
@@ -320,7 +321,13 @@ class Cache {
       } on FileSystemException {
         if (!printed) {
           _logger.printTrace('Waiting to be able to obtain lock of Flutter binary artifacts directory: ${_lock!.path}');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/cache.dart:323]: shouldn't this go to stderr too?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart:323]: oh, I guess you wouldn't be capturing output in verbose mode, nvm

**Comment by christopherfujino** [comment]
on Windows:

```
21:53 +47 ~2 -1: test\integration.shard\cache_test.dart: Cache.lock should log a message to stderr when lock is not acquired [E]                                                                       
  Bad state: No element
  dart:core                                      List.single
  test\integration.shard\cache_test.dart 128:28  main.<fn>.<fn>
```

**Comment by gspencergoog** [comment]
Yeah, I was looking at that.  It seems like Windows might be flakier than the other platforms when it comes to locking performance.  If I step through in the debugger, the test passes, but timing prevents it from passing when run normally.  I might just disable this test on Windows: the test itself is just to make sure we're passing things to stderr, and we don't necessarily verify that on all platforms (although it would be nice).

File: packages/flutter_tools/test/integration.shard/cache_test.dart [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import 'package:flutter_tools/src/base/logger.dart';
+import 'package:flutter_tools/src/base/terminal.dart';
+import 'package:flutter_tools/src/cache.dart';
+import 'package:process/process.dart';
+
+import '../src/common.dart';
+import '../src/context.dart';
+import 'test_utils.dart';
+
+class FakeLogger extends Logger {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/integration.shard/cache_test.dart:21]: Please use the BufferLogger class instead

Comment (gspencergoog) [reply packages/flutter_tools/test/integration.shard/cache_test.dart:21]: Ahh, that's what it's called.  I was looking for something like that (expected it to be called "Fake" something). OK.

**Comment by christopherfujino** [comment]
@gspencergoog bad merge?

**Comment by gspencergoog** [comment]
Yep.  Fixed.

File: packages/flutter_tools/test/integration.shard/cache_test.dart (hunk 2 above)
Comment (gspencergoog) [reply packages/flutter_tools/test/integration.shard/cache_test.dart]: OK, Fixed.  PTAL

**Comment by gspencergoog** [comment]
@jonahwilliams I think I resolved everything, please take another look.

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Pull request merged**